File: common-lib/src/com/bigcorp/common/rules/impl/ShortSaleRule.java
package com.bigcorp.common.rules.impl;

import com.bigcorp.common.billing.CommissionCalculator;
import com.bigcorp.common.rules.Rule;
import com.bigcorp.common.rules.RuleContext;
import com.bigcorp.common.model.TradeOrder;

/**
 * Short sale restriction — rejects SELL orders above 1000 shares.
 * 
 * Added after compliance flagged large short sales in Q3 2003.
 * The 1000 share threshold was "what felt right" per the trading desk.
 * 
 * @author feature-rusher
 * @since 2003 Q4
 */
public class ShortSaleRule implements Rule {

    // Resolved: was copy-pasted constant, now uses centralized CommissionCalculator (JIRA-2501)

    public String getName() {
        return "ShortSale";
    }

    public int getPriority() {
        return 75; // after MarketHours(80), before SpecialClients(50)
    }

    public boolean isActive() {
        return true;
    }

    public boolean evaluate(RuleContext context) {
        try {
            TradeOrder order = context.getOrder();

            if (TradeOrder.SIDE_SELL.equals(order.getSide())) {
                int quantity = order.getQuantity();

                if (quantity > 1000) {
                    context.reject("Short sale limit exceeded");
                    return false;
                }

                // stash commission for downstream — uses centralized rate from client tier
                String tier = (context.getClient() != null) ? context.getClient().getTier() : null;
                double rate = CommissionCalculator.getRate(tier);
                context.setAttribute("short_sale_commission", Double.valueOf(quantity * rate));
            }

            return true;
        } catch (Exception e) {
            System.err.println("ShortSaleRule error: " + e.getMessage());
            return true;
        }
    }

    public void execute(RuleContext context) {
        // nothing to do
    }
}

File: rewrite-experiment/hypothesis-a/src/rules/ShortSaleRule.ts
import { Rule } from './Rule';
import { RuleContext } from '../domain/RuleContext';
import { OrderSide } from '../domain/enums';
import { getCommissionRate } from '../pricing/CommissionCalculator';
import { ClientTier } from '../domain/enums';

const SHORT_SALE_LIMIT = 1000;

export class ShortSaleRule implements Rule {
  readonly name = 'ShortSaleRule';
  readonly priority = 75;
  readonly category = 'Business';

  isActive(): boolean {
    return true;
  }

  async evaluate(ctx: RuleContext): Promise<boolean> {
    const { order } = ctx;
    if (!order) return true;

    if (order.side === OrderSide.SELL && order.quantity > SHORT_SALE_LIMIT) {
      ctx.reject('Short sale limit exceeded');
      return false;
    }

    return true;
  }

  async execute(ctx: RuleContext): Promise<void> {
    const { order, client } = ctx;
    if (!order || order.side !== OrderSide.SELL) return;

    const tier = client?.tier as ClientTier | undefined;
    const rate = getCommissionRate(tier);
    ctx.setAttribute('commission_rate', rate);
  }
}
